api/src/test/java/io/grpc/MetadataTest.java

```diff
@@ -334,6 +390,96 @@ public void invalidKeyName() {
     }
   }
 
+  @Test
+  public void getSameKeyReturnsSameInstnce() {
+    Fish salmon = new Fish("salmon");
+    Metadata h = new Metadata();
+    h.put(KEY_STREAMED, salmon);
+    assertSame(salmon, h.get(KEY_STREAMED));
```

Comment:
Is this a guarantee we actually want to provide? I was thinking we were saying we'd let the InputStream hacks take care of this, so objects would be duplicated by default and mutable objects would be "safe."

I think you were making use of this in your example consuming code internally. Getting rid of this would require using the same InputStream trick that's used for outbound messages?

Reply:
Good point, this was effectively testing the implementation, changed to assertEquals (also above).